APIView code file token tsp and json schema

praveenkuttappan opened the pull request:
Here is the token schema based on the discussion with @tjprescott and @JonathanGiles. This schema is closer to older token file but at the same time we will have parent child relationship between the code lines to have the context details for APIView. Some other changes are to remove new line and white space tokens as well as to generate navigation TOC by APIView so parser does not haver to generate it.

praveenkuttappan commented:
I will submit a separate PR to remove https://github.com/Azure/azure-sdk-tools/tree/main/tools/apiview/parsers/apiview-treestyle-parser-schema once new token model is signed off.

maririos commented on tools/apiview/token-schema/model.tsp:
I liked the location of the previous file. https://github.com/Azure/azure-sdk-tools/tree/main/tools/apiview/parsers 
that way it is next to the other parsers and we could use the [contributing](https://github.com/Azure/azure-sdk-tools/blob/main/tools/apiview/parsers/CONTRIBUTING.md) to explain how to consume it, test, etc

praveenkuttappan replied:
token schema is also consumed by TypeSpec emitter which is aa sibling of `parsers`

maririos replied:
> token schema is also consumed by TypeSpec emitter which is aa sibling of `parsers`

Still don't see this as a problem. the purpose of this file is to be able to communicate the API with the parsers. It would make more sense to have it there too. Also not a hill I will die on :)

praveenkuttappan replied:
Done.

maririos commented:
Can we generate examples too to see how it will look?

maririos commented on tools/apiview/token-schema/model.tsp:
```diff
@@ -0,0 +1,89 @@
+   */
+  navigateToId?: string;
+  /** set skipDiff to true if underlying token needs to be ignored from diff calculation. For e.g. package metadata or dependency versions 
+   *  are usually not required to be excluded when comparing two revisions to avoid reporting them as API changes*/
```
```suggestion
   *  are usually not required so they can be excluded when comparing two revisions to avoid reporting them as API changes*/
```

praveenkuttappan commented:
> Can we generate examples too to see how it will look?

As @lmazuel mentioned on another PR, I am trying to use TypeSpec C# emitter to directly generate the .NET models from TypeSpec and then I will create sample token file for Template. We can include this sample file along with the schema for comparison.

tjprescott commented on tools/apiview/token-schema/model.tsp:
```diff
@@ -0,0 +1,107 @@
+  /** Set isDocumentation to true if current token is part of documentation */
+  isDocumentation?: boolean = false;
+  /** Language specific style css class names */
+  languageStyleClasses?: Array<string>;
```
Assuming this is the equivalent of `RenderClasses`?

praveenkuttappan replied:
yes. I will rename it.

tjprescott commented on tools/apiview/token-schema/model.tsp:
```diff
@@ -0,0 +1,107 @@
+   *  are usually excluded when comparing two revisions to avoid reporting them as API changes*/
+  skipDiff?: boolean = false;
+  /** This is set if API is marked as hidden */
+  isHidden?: boolean = false;
```
Why is "isHidden" on the token rather than the line? I ask because I'm thinking about how to recreate the TypeSpec navigation. I would have some top-level CodeLines for Resources, Models, etc. but these should not appear in the actual text of the APIView. They _will_ have children. So it seems to me that the CodeLine should be hidden (or not), and if the line is hidden, it will be important to factor that into the automatic indentation. 

Are there use cases for hidden tokens within a line the way there was for hiding things from the diff?

praveenkuttappan replied:
I don't see any special use case. I can move it to line model.

praveenkuttappan commented on tools/apiview/token-schema/model.tsp:
```diff
@@ -0,0 +1,107 @@
+  crossLanguagePackageId?: string;
+  codeLines: Array<CodeLine>;
+  /** Add any system generated comments for the current token. */
+  systemComments?: Array<CodeDiagnostic>;
```
I will update the docstring and property names as well. diagnostic will be at code file level as it used to and you can link it to line ID.

maririos commented:
> > Can we generate examples too to see how it will look?
> 
> @maririos @tjprescott I have added a sample token file generated for .NET Azure.Template package and rendered review content text file from this sample token(without docs and system generated comments).

This is great! we should improve then the documentation on https://github.com/Azure/azure-sdk-tools/blob/main/tools/apiview/parsers/CONTRIBUTING.md with this info and how to use the json schema to validate the parser files

praveenkuttappan commented:
> > > Can we generate examples too to see how it will look?
> > 
> > 
> > @maririos @tjprescott I have added a sample token file generated for .NET Azure.Template package and rendered review content text file from this sample token(without docs and system generated comments).
> 
> This is great! we should improve then the documentation on https://github.com/Azure/azure-sdk-tools/blob/main/tools/apiview/parsers/CONTRIBUTING.md with this info and how to use the json schema to validate the parser files

Updated contributing guide and info about json schema validator.

maririos commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -1,1165 +1,44 @@
 - Faster diffing using tree shaker instead of current text based comparison.
 - Provide diffing with context of where the line that changed belong to in the tree, instead of showing the 5 lines before and after the change.
 - Provide cross language view for a granular section within an API review.
-- Support for collapsible views at each class and namespace level to make it easier to review the change. User will be able to expand/collapse an entire class or a namespace and it’s children.
```
you are removing this because it is not yet implemented or because with the proposed changes it won't be possible

praveenkuttappan replied:
> you are removing this because it is not yet implemented or because with the proposed changes it won't be possible

It's not added yet. I am removing it since it's not a benefit yet straight from this new token model. We can add it once the feature is added in APIView.

maririos commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -1,1165 +1,44 @@
 ## Key concepts
 
-Previously APIview tokens were created as a flat list assigned to the `CodeFileToken[] Tokens`  property of the [CodeFile](../../../src/dotnet/APIView/APIView/Model/CodeFile.cs). Then the page navigation is created and assigned to `NavigationItem[] Navigation`. For tree style tokens these two properties are no longer required, instead a `List<APITreeNode> APIForest` property will be used to captu ...
+APIView token schema is available in [TypeSpec](./apiview-treestyle-parser-schema/model.tsp) and [JSON](./apiview-treestyle-parser-schema/model.json). Parser needs to create a `CodeFile` object as per the schema.
```
maybe?
```suggestion
APIView token schema is available in [TypeSpec](./apiview-treestyle-parser-schema/model.tsp) and [JSON](./apiview-treestyle-parser-schema/model.json). Language parsers needs to create a `CodeFile` object as per the schema.
```

maririos commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -1,1165 +1,44 @@
 
-Previously APIview tokens were created as a flat list assigned to the `CodeFileToken[] Tokens`  property of the [CodeFile](../../../src/dotnet/APIView/APIView/Model/CodeFile.cs). Then the page navigation is created and assigned to `NavigationItem[] Navigation`. For tree style tokens these two properties are no longer required, instead a `List<APITreeNode> APIForest` property will be used to captu ...
+APIView token schema is available in [TypeSpec](./apiview-treestyle-parser-schema/model.tsp) and [JSON](./apiview-treestyle-parser-schema/model.json). Parser needs to create a `CodeFile` object as per the schema.
+`CodeFile` object contains a metadata about the package and array of `ReviewLine` object. Each object of review line in `CodeFile` object is a top level lines to be listed. For e.g. Top level lines are mostly namespace or module level nodes.
```
```suggestion
`CodeFile` object contains metadata about the package and an array of `ReviewLine` objects. Each object of `ReviewLine` in `CodeFile` object is a top-level line to be listed. For e.g. Top-level lines are mostly namespace or module level nodes.
```

maririos commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -1,1165 +1,44 @@
-- Each tree node has top tokens which should be used to capture the main tokens on the node, these can span multiple lines.
-- Module name, decorators, and parameters should be modeled as `TopTokens`.
-- If the language requires it use the bottom tokens to capture tokens that closes out the node, this is usually just the closing bracket and/or empty lines.
+APIView generates navigation tree based on the information in token. A token is included in the navigation tree if `NavigatonDisplayName` is set in `ReviewToken` and `LineId` is set in `ReviewLine` object that contains the `ReviewToken`
```
```suggestion
APIView generates a navigation tree based on the information in the tokens. A token is included in the navigation tree if `NavigatonDisplayName` is set in `ReviewToken` and `LineId` is set in `ReviewLine` object that contains the `ReviewToken`
```

maririos commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -1,1165 +1,44 @@
-Serialize the generated code file to JSON with Gzip compression. The output file should have `.json.tgz` extension. Try to make the json as small as possible by ignoring null values and empty collections.
-Don't worry about indentation that will be handled by the tree structure. In the case you want to have indentation between the tokens then use `TabSpace` token kind.
+Serialize the generated code file to JSON. The output file should have `.json` extension. Try to make the json as small as possible by ignoring null values and empty collections.
+Don't worry about indentation that will be handled by the tree structure based on the parent - child relationship among `ReviewLine` objects.
```
```suggestion
Don't worry about indentation as it will be handled by the tree structure based on the parent-child relationship among `ReviewLine` objects.
```

maririos commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -1,1165 +1,44 @@
-<details>
-<summary>click to expand json</summary>
+- Select `Custom` as schema type and copy and paste the contents in [json schema](./apiview-treestyle-parser-schema/model.json) to left panel.
+- Generate API view token file and paste generated JSON content onto right side panel to validate.
```
```suggestion
- Generate APIView token file and paste generated JSON content onto right side panel to validate.
```

maririos commented on tools/apiview/parsers/CONTRIBUTING.md:
```diff
@@ -1,1165 +1,44 @@
-Make a token a Url by adding key `NavigateToUrl` to the token property and setting the value to the target Url.
-<details>
-<summary>click to expand json</summary>
+- Select `Custom` as schema type and copy and paste the contents in [json schema](./apiview-treestyle-parser-schema/model.json) to left panel.
```
Jonathan was saying we can publish the json schema on https://www.schemastore.org/json/
or somewhere and that way people can just reference and there is no need for them to manually copy it every time they want to validate it.
Look in Teams for the `APIView Documentation` group chat. Jonathan sent the message on 7/23

praveenkuttappan replied:
I have discussed this with Jonathan and he is fine with JSON schema stored in azure-sdk-tools repo.

maririos commented on tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:
```diff
@@ -0,0 +1,94 @@
+  PackageVersion: string;
+  /** version of the APIview language parser used to create token file*/
+  ParserVersion: string;
+  Language: "C"|"C++"|"C#"|"Go"|"Java"|"JavaScript"|"Kotlin"|"Python"|"Swagger"|"Swift"|"TypeSpec";
```
For my own curiosity and my lack of knowledge here 😨 , we support Java with variant Android, and also Kotlin? do people use both? or how does it work?

praveenkuttappan replied:
It's handled by Java parser.

maririos commented on tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:
```diff
@@ -0,0 +1,94 @@
+model ReviewLine {
+  /** lineId is only required if we need to support commenting on a line that contains this token. 
+   *  Usually code line for documentation or just punctuation is not required to have lineId. lineId should be a unique value within 
+   *  the review token file to use it assign to review comments as well as navigation Id within the review page.
```
wording here seems strange. is the `lineID` assigned to review comment and navigation ID? or is it use to compare against those?

praveenkuttappan replied:
Line ID is not used for comparison. It's only used to identify a line from a given Line ID so comments use it to identify where a comment needs to be listed. Comment stores this Line ID in DB. Similarly, Line ID is used as HREF within the review so a type name can navigate to its type definition.

maririos commented on tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:
```diff
@@ -0,0 +1,94 @@
+  /** Add any child lines as children. For e.g. all classes and namespace level methods are added as a children of namespace(module) level code line. 
+   *  Similarly all method level code lines are added as children of it's class code line.*/
+  Children?: Array<ReviewLine>;
+  /** Set current line as hidden code line by default. .NET has hidden APIs and architects does not want to see them by default. */
```
```suggestion
  /** Set current line as hidden code line by default. .NET has hidden APIs and architects don't want to see them by default. */
```

maririos commented on tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:
```diff
@@ -0,0 +1,94 @@
+
+/** System comment object is to add system generated comment. It can be one of the 4 different types of system comments. */
+model CodeDiagnostic {
+  //Diagnostic ID is auto generated ID by CSharp analyzer.
```
```suggestion
  /** Diagnostic ID is auto generated ID by CSharp analyzer. */
```

maririos commented on tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:
```diff
@@ -0,0 +1,94 @@
+  DiagnosticId?: string;
+    /** Id of ReviewLine object where this diagnostic needs to be displayed(Options*/
+  TargetId: string;
+  // Auto generated system comment to be displayed under targetted line.
```
```suggestion
  /** Auto generated system comment to be displayed under targeted line. */
```

maririos commented on tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:
```diff
@@ -0,0 +1,94 @@
+model CodeDiagnostic {
+  //Diagnostic ID is auto generated ID by CSharp analyzer.
+  DiagnosticId?: string;
+    /** Id of ReviewLine object where this diagnostic needs to be displayed(Options*/
```
```suggestion
    /** Id of ReviewLine object where this diagnostic needs to be displayed */
```

maririos commented on tools/apiview/parsers/apiview-treestyle-parser-schema/model.tsp:
```diff
@@ -0,0 +1,94 @@
+ Info: 1,
+ Warning: 2,
+ Error:	3,
+ Fatal:	4
```
it might be good to add the documentation that says that fatal will block an API from being approved unless strictly overwritten by approves
